File: crates/gateway-auth-codex/src/revoke.rs
#![forbid(unsafe_code)]

use crate::auth_json::AuthJson;
use crate::auth_json::Tokens;
use crate::{CodexAuthError, load_auth_json};
use serde::Serialize;
use std::path::Path;
use std::time::Duration;

const REVOKE_TOKEN_URL: &str = "https://auth.openai.com/oauth/revoke";
const REVOKE_HTTP_TIMEOUT: Duration = Duration::from_secs(10);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum RevokeTokenKind {
    Access,
    Refresh,
}

impl RevokeTokenKind {
    fn as_str(self) -> &'static str {
        match self {
            Self::Access => "access_token",
            Self::Refresh => "refresh_token",
        }
    }

    fn client_id(self) -> Option<&'static str> {
        match self {
            Self::Access => None,
            Self::Refresh => Some("app_EMoamEEZ73f0CkXaXp7hrann"),
        }
    }
}

#[derive(Serialize)]
struct RevokeTokenRequest<'a> {
    token: &'a str,
    token_type_hint: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    client_id: Option<&'static str>,
}

pub fn logout(path: &Path) -> Result<bool, CodexAuthError> {
    match std::fs::remove_file(path) {
        Ok(()) => Ok(true),
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Ok(false),
        Err(err) => Err(CodexAuthError::IoWithPath {
            path: path.to_path_buf(),
            source: err,
        }),
    }
}

pub async fn logout_with_revoke(path: &Path) -> Result<bool, CodexAuthError> {
    let auth = match load_auth_json(path) {
        Ok(auth) => auth,
        Err(CodexAuthError::AuthNotFound { .. }) => return logout(path),
        Err(err) => return Err(err),
    };

    if let Some((token, kind)) = revocable_token(&auth)
        && let Err(err) = revoke_oauth_token(REVOKE_TOKEN_URL, token, kind).await
    {
        tracing::warn!("failed to revoke auth tokens during logout: {err}");
    }

    logout(path)
}

fn revocable_token(auth: &AuthJson) -> Option<(&str, RevokeTokenKind)> {
    let tokens: &Tokens = auth.tokens.as_ref()?;
    if let Some(refresh) = tokens
        .refresh_token
        .as_deref()
        .filter(|value| !value.is_empty())
    {
        Some((refresh, RevokeTokenKind::Refresh))
    } else {
        tokens
            .access_token
            .as_deref()
            .filter(|value| !value.is_empty())
            .map(|token| (token, RevokeTokenKind::Access))
    }
}

async fn revoke_oauth_token(
    endpoint: &str,
    token: &str,
    kind: RevokeTokenKind,
) -> Result<(), std::io::Error> {
    let request = RevokeTokenRequest {
        token,
        token_type_hint: kind.as_str(),
        client_id: kind.client_id(),
    };

    let client = reqwest::Client::new();
    let response = client
        .post(endpoint)
        .header("Content-Type", "application/json")
        .timeout(REVOKE_HTTP_TIMEOUT)
        .json(&request)
        .send()
        .await
        .map_err(std::io::Error::other)?;

    let status = response.status();
    if status.is_success() {
        Ok(())
    } else {
        let body = response.text().await.unwrap_or_default();
        Err(std::io::Error::other(format!(
            "failed to revoke {}: {}: {}",
            kind.as_str(),
            status,
            body
        )))
    }
}

File: core/impl/port/auth/codexauth/revoke.go
package codexauth

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"

	"github.com/Bytonomics/cld-gateway/netpolicy"
)

const (
	revokeTokenURL    = "https://auth.openai.com/oauth/revoke"
	revokeHTTPTimeout = 10 * time.Second
	revokeRefreshCID  = "app_EMoamEEZ73f0CkXaXp7hrann"
)

type revokeTokenKind int

const (
	revokeAccessToken revokeTokenKind = iota
	revokeRefreshToken
)

func (k revokeTokenKind) hint() string {
	switch k {
	case revokeRefreshToken:
		return "refresh_token"
	default:
		return "access_token"
	}
}

func (k revokeTokenKind) clientID() string {
	if k == revokeRefreshToken {
		return revokeRefreshCID
	}
	return ""
}

type revokeTokenRequest struct {
	Token         string `json:"token"`
	TokenTypeHint string `json:"token_type_hint"`
	ClientID      string `json:"client_id,omitempty"`
}

// revocableToken ports revoke::revocable_token: prefers a non-empty refresh
// token, falling back to a non-empty access token.
func revocableToken(t *tokens) (token string, kind revokeTokenKind, ok bool) {
	if t == nil {
		return "", 0, false
	}
	if t.RefreshToken != nil && *t.RefreshToken != "" {
		return *t.RefreshToken, revokeRefreshToken, true
	}
	if t.AccessToken != nil && *t.AccessToken != "" {
		return *t.AccessToken, revokeAccessToken, true
	}
	return "", 0, false
}

// revokeOAuthToken ports revoke::revoke_oauth_token.
func revokeOAuthToken(ctx context.Context, httpClient *netpolicy.Client, endpoint, token string, kind revokeTokenKind) error {
	payload := revokeTokenRequest{
		Token:         token,
		TokenTypeHint: kind.hint(),
		ClientID:      kind.clientID(),
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	ctx, cancel := context.WithTimeout(ctx, revokeHTTPTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	res, err := httpClient.Do(req)
	if err != nil {
		return err
	}
	defer func() { _ = res.Body.Close() }()

	if res.StatusCode >= 200 && res.StatusCode < 300 {
		return nil
	}
	respBody, _ := io.ReadAll(res.Body)
	return fmt.Errorf("failed to revoke %s: %s: %s", kind.hint(), res.Status, string(respBody))
}

// revokeTokens ports logout_with_revoke's revocation step: best-effort
// revocation of the account's refresh (preferred) or access token. Unlike
// the Rust implementation, which silently discards revocation failures via
// tracing::warn, this logs the error through slog so cleanup failures remain
// observable (approved gap fix G5).
func (s *Store) revokeTokens(ctx context.Context) {
	parsed, err := s.loadAuthJSON()
	if err != nil {
		if !errors.Is(err, ErrAuthNotFound) {
			slog.Warn("failed to load auth.json for token revocation", "error", err)
		}
		return
	}

	token, kind, ok := revocableToken(parsed.Tokens)
	if !ok {
		return
	}

	if err := revokeOAuthToken(ctx, s.httpClient(), revokeTokenURL, token, kind); err != nil {
		slog.Warn("failed to revoke auth tokens during logout", "error", err)
	}
}
